torch/distributed/checkpoint/state_dict_saver.py

```diff
@@ -212,9 +215,58 @@ def async_save(
-
-    executor = ThreadPoolExecutor(max_workers=1)
+    # This is the simplest implementation of a "stage" specific method which occurs pre-execution
+    # This is done in a way that is BC, and does not force the creation of `stage_state_dict` for
+    # users that don't really want to think about this
+
+    # Issue #1
+    # `staging` still has no explicit path to the same local/global planning
+    # via the usual planner interface. If we want to do things like dedup replicas to avoid
+    # extraneous D2H pre- submitting the save job, we still don't have an (explicit) way of handling that
+    # in this proposal. (Users could potentially call `dcp.save inside their custom 'stage' method`)
+
+    # Issue #2
+    # Users are forced to use a thread as the async vehicle, or provide some ThreadPoolExecutor-like object
+    # Is there a cleaner way to implement this without relying on ThreadPoolExecutor? Do we want to support this?
+    # This may not be true for users which hack storage_writer and planner methods to be async under the hood, but
+    # IMO the UX is shaky.
+    # This might be important in xlformers (eventually) or in cases
+    # where users are worried about GIL issues.
+
+    # A totally reasonable alternative to this entire proposal is to expect users who want extra
+    # customization to implement their own async_save methods, and rely on internal calls to dcp.save
+    # in fact even with these extensions, this might still be what we tell users to do if this isn't
+    # "good enough"
+    # here's once example which accomplishes this and addresses both issues
+
+
+    # def custom_async_save(...):
+    #     this step accomplishes staging and includes the usual 'planning' calls (issue 1)
+    #     buffered_writer = CpuBufferedWriter() # this is stateful, contains a copy of state_dict
+    #     dcp.save(state_dict, storage_writer=buffered_writer)
+
+    #     final_storage_writer = FileSystemWriter()
+    #     create whatever thread you want (issue #2)
+    #     mp.spawn(
+    #         dcp.save,    # or some custom sub-process method wich calls dcp.save under the hood
+    #         save,
+    #         buffered_writer.state_dict,   # lot's of way's to do this, not really the most important part
+    #         checkpoint_id=checkpoint_id,
+    #         storage_writer=storage_writer,
+    #         planner=planner,
+    #         process_group=process_group,
+    #     )
+    # leaving out the rest of the details for managing your extra special subprocess.
+
+    # another thing I don't like about this is passing state_dict to storage writer, returning it out, and then passing
+    # it back in eventually in the subsequent `save_state_dict` call (it goes async_save -> save -> _save_state_dict)
+    # another fix could be to make the storage_writer handle this in a more `stateful` way
+    if isinstance(storage_writer, AsyncStager):
+        cpu_state_dict = storage_writer.stage(state_dict)
```

Comment:
Is there a restriction on the stage function having to be purely local? In practice, can it do some cross-rank synchronization in the beginning, e.g. some validations, available memory budget calculation aware of max tensor sizes on each of the local ranks, or even in the future a common memory budget for the current trainer host.

Reply:
Could this be handled in `stage` by doing something like `dist.get_rank()==0` and returning early? 

This is essential for a lot of cases where we do want local staging to occur on every rank. That seems the more common path to me

Reply:
I may have misunderstood -- in this proposal `stage` runs in every rank. If collectives are necessary, my best guess is they are implemented inside the `stage` method

Reply:
> I may have misunderstood -- in this proposal `stage` runs in every rank. If collectives are necessary, my best guess is they are implemented inside the `stage` method

Sound good!